### PR TITLE
gh-121249: Support _Complex types in the struct module

### PR DESCRIPTION
**Pull request by skirpichev**
* Issue: gh-121249

----
📚 Documentation preview 📚: https://cpython-previews--121613.org.readthedocs.build/

### CODE REVIEW

File: Doc/library/struct.rst
Comment (ZeroIntensity) [inline Doc/library/struct.rst]: This should probably link to the specification.

File: Doc/library/struct.rst
Comment (ZeroIntensity) [inline Doc/library/struct.rst:368]: Again, link to the spec, we shouldn't expect people to have arbitrary specification numbers memorized.

File: Misc/NEWS.d/next/Library/2024-07-10-08-13-34.gh-issue-121249.W9Gd09.rst
Comment (ZeroIntensity) [inline Misc/NEWS.d/next/Library/2024-07-10-08-13-34.gh-issue-121249.W9Gd09.rst]:
```suggestion
C types in the :mod:`struct` module if the compiler has C11 complex
```

File: Misc/NEWS.d/next/Library/2024-07-10-08-13-34.gh-issue-121249.W9Gd09.rst
Comment (ZeroIntensity) [inline Misc/NEWS.d/next/Library/2024-07-10-08-13-34.gh-issue-121249.W9Gd09.rst]:
```suggestion
Support the :c:expr:`float complex` and :c:expr:`double complex`
```

File: Modules/_struct.c
Comment (ZeroIntensity) [inline Modules/_struct.c:1]: It's probably a better idea to define these functions even when `Py_HAVE_C_COMPLEX` isn't enabled, and then just raise a `struct.error` indicating that the system doesn't support it. I could see some people in the future being very confused as to why `struct` is complaining about an invalid format string across different systems.

**Review by ZeroIntensity — CHANGES_REQUESTED** [review]
Nice PR! This looks like something that would have had to get approved on Discourse, though.

File: Doc/library/struct.rst
Comment (skirpichev) [reply Doc/library/struct.rst]: Which link do you suggest?  Wikipedia page for IEEE 754, C11 standard?

File: Doc/library/struct.rst
Comment (skirpichev) [reply Doc/library/struct.rst:368]: I doubt it's arbitrary.  We have many references to the IEEE 754 standard in docs.

**Review by skirpichev — COMMENTED** [review]
> It's probably a better idea to define these functions even when Py_HAVE_C_COMPLEX isn't enabled, and then just raise a struct.error indicating that the system doesn't support it.

Patch follows to the ctypes module.  Not sure, maybe your approach is better.  Lets leave this to core devs.

**Comment by ZeroIntensity** [comment]
> Lets leave this to core devs.

This is why things get discussed on discourse first :wink:

File: Doc/library/struct.rst
Comment (ZeroIntensity) [reply Doc/library/struct.rst]: Probably the C11 standard.

File: Doc/library/struct.rst
Comment (skirpichev) [inline Doc/library/struct.rst]:
```suggestion
Additionally, if IEC 60559 compatible complex arithmetic (Annex G of the
C11 standard) is supported, the following format characters are available:
```
?

File: Lib/test/test_struct.py
Comment (skirpichev) [inline Lib/test/test_struct.py]: These helpers refactored to a mixin class in https://github.com/python/cpython/pull/121071

File: Doc/library/struct.rst
Comment (ZeroIntensity) [reply Doc/library/struct.rst]: That would be perfect.

**Review by ZeroIntensity — APPROVED** [review]
This implementation LGTM.

Although, for core dev review, I'm not too sure about this behavior: if the compiler does not support C11 complex types, then the new format characters are disabled and raise an error with no further information.

I'm worried that this will lead to some bug reports later on wondering why `E` and `C` don't work on their system -- it would be ideal to raise a different error mentioning that the system doesn't support C11 (albeit, it would be more work to implement).

File: Doc/library/struct.rst
Comment (picnixz) [inline Doc/library/struct.rst]:
```suggestion
   Added support for the ``'E'`` and ``'C'`` formats.
```

To be consistent with the previous 'versionchanged' entries.

File: Doc/library/struct.rst
Comment (picnixz) [inline Doc/library/struct.rst]:
```suggestion
   For the ``'E'`` and ``'C'`` format characters, the packed representation uses
```

I would use the same terminology as in the other notes.

**Review by picnixz — COMMENTED** [review]
Some minor comments but I also agree that having an error without info for an unknown format if complex are not supported is a bit weird.

**Comment by skirpichev** [comment]
Now on systems without _Complex type - a different error will be raised.

File: Modules/_struct.c
Comment (skirpichev) [reply Modules/_struct.c:1]: Done

File: Lib/test/test_struct.py
Comment (picnixz) [inline Lib/test/test_struct.py]: I would make two separate tests and use `skipIf` depending on the `have_c_complex` value. By the way, which build bots do not support float/double complex?

Comment (skirpichev) [reply Lib/test/test_struct.py]: > which build bots do not support float/double complex?

E.g. on Windows, WASI.

**Review by picnixz — APPROVED** [review]
On my side it's fine. I didn't check in details the functions bodies, but I assume that the tests should cover most of the cases. I think it remains to wait for Victor's and/or Serhiy's review.

**Comment by skirpichev** [comment]
As it's a follow up of https://github.com/python/cpython/pull/120894, CC @vstinner

File: Lib/test/test_struct.py
Comment (vstinner) [inline Lib/test/test_struct.py]:
```suggestion
    @unittest.skipUnless(have_c_complex, "requires C11 complex type support")
```

File: Lib/test/test_struct.py
Comment (vstinner) [inline Lib/test/test_struct.py]:
```suggestion
    @unittest.skipIf(have_c_complex, "requires no C11 complex type support")
```

File: Modules/_struct.c
Comment (vstinner) [inline Modules/_struct.c]:
```suggestion
#  define FLOAT_COMPLEX_ALIGN (sizeof(st_float_complex) - sizeof(float complex))
#  define DOUBLE_COMPLEX_ALIGN (sizeof(st_double_complex) - sizeof(double complex))
```

File: Modules/_struct.c
Comment (vstinner) [inline Modules/_struct.c]: I'm not sure that the `(char *)` cast is useful. Same remark below.

File: Modules/_struct.c
Comment (vstinner) [inline Modules/_struct.c]:
```suggestion
    double x = PyFloat_Unpack4(p, 0);
    if (x == -1.0 && PyErr_Occurred()) {
        return NULL;
    }
    double y = PyFloat_Unpack4(p + 4, 0);
    if (y == -1.0 && PyErr_Occurred()) {
        return NULL;
    }
```

Same suggestion below.

File: Modules/_struct.c
Comment (vstinner) [inline Modules/_struct.c]:
```suggestion
    double x = PyFloat_Unpack4(p, 1);
    if (x == -1.0 && PyErr_Occurred()) {
        return NULL;
    }
    double y = PyFloat_Unpack4(p + 4, 1);
    if (y == -1.0 && PyErr_Occurred()) {
        return NULL;
    }
```

Same suggestion below.

File: Modules/_struct.c
Comment (vstinner) [inline Modules/_struct.c]: It shouldn't be p+8 here? If it's a real typo, I'm curious why it was missed by tests.

File: Modules/_struct.c
Comment (skirpichev) [reply Modules/_struct.c]: I suspect it's useless as well.  Yet, it's used in all helpers, e.g. np_double right above.

Maybe clean up these casts across the file?

File: Modules/_struct.c
Comment (skirpichev) [reply Modules/_struct.c]: I would guess that CI has only little-endian machines.  np_double_complex() was working instead.  Typo fixed.

File: Modules/_struct.c
Comment (skirpichev) [reply Modules/_struct.c]: From the statistics, it seems that such remnants are mostly in this module:
```
$ git grep 'memcpy.*(char *\*)' | cut -f1 -d:|sort -n|uniq -c|sort -n
      1 Include/internal/pycore_call.h
      1 Modules/_ctypes/cfield.c
      1 Objects/call.c
      1 Objects/longobject.c
      1 Python/dtoa.c
      2 Modules/_io/bufferedio.c
      2 Objects/unicodeobject.c
      3 Modules/socketmodule.c
      5 Objects/memoryobject.c
     30 Modules/_struct.c
```

File: Modules/_struct.c
Comment (vstinner) [reply Modules/_struct.c]: Oh right, `_structmodule_exec()` modifies `lilendian_table` at startup to use the native endian implementation instead. A big endian buildbot would catch the typo.

**Review by vstinner — APPROVED** [review]
LGTM

**Comment by vstinner** [comment]
Merged, thanks.

**Pull request merged**
